### DISCUSSION TITLE
dts: How to manually run the C preprocessor on DTS files?

### DISCUSSION THREAD

**Discussion by soares-sergio** [zephyrproject-rtos/zephyr#66721] [Q&A]
Hi folks,

Context: I'm trying to learn more about DTS and how it can go from `.dts` to `devicetree.h`. I'm working in a non-Zephyr codebase and I'm investigating if it would be possible to reuse Zephyr's DTS tools/bindings so we don't have to duplicate some of this code in our own codebase.

To start, I'm using the `mimxrt595_evk_cm33.dts` as an example and trying to manually run the steps to convert it into a header file.

1. The first step is to run the C preprocessor on the DTS file to resolve macros and includes. I'm running the `cpp` command directly and providing include directories manually:
```
cpp zephyr/boards/arm/mimxrt595_evk/mimxrt595_evk_cm33.dts -Izephyr/include -Izephyr/dts -Izephyr/dts/arm -Izephyr/dts/common -Imodules/hal/nxp/dts -traditional-cpp -o mimxrt595_evk_cm33_preprocessed.dts
```

It seems to run mostly fine but at the end I get multiple errors such as:
```
zephyr/include/zephyr/sys/util_macro.h:649: error: syntax error in macro parameter list
  649 | #define MACRO_MAP_CAT_N(N, ...) MACRO_MAP_CAT_N_(N, __VA_ARGS__)
```

Note: I had to add `-traditional-cpp` otherwise it would fail on specific macros such as `#address-cells` etc.
My cpp version is `13.2.0`.

Any ideas on what could be wrong? Thanks!

**Comment by yashi** [comment]
`cmake/modules/dts.cmake` is the driving functions.  In that, you see:

```cmake
set(GEN_DEFINES_SCRIPT          ${DT_SCRIPTS}/gen_defines.py)
```

which is `scripts/dts/gen_defines.py`.  At the top of the script you read

```python
# This script uses edtlib to generate a header file from a devicetree
# (.dts) file. Information from binding files in YAML format is used
# as well.

```
